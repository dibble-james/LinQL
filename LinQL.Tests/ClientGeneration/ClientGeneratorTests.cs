namespace LinQL.Tests.ClientGeneration;

using System.Text;
using LinQL.ClientGeneration;
using Xunit;

public class ClientGeneratorTests
{
    [Fact]
    public void StarWarsClient()
        => Snapshot.Match(ClientGenerator.Generate(
            Encoding.UTF8.GetString(SDLs.StarWars),
            "StarWars",
            Array.Empty<string>()));

    [Fact]
    public void ShiftshareClient()
        => Snapshot.Match(ClientGenerator.Generate(
            Encoding.UTF8.GetString(SDLs.Shiftshare),
            "Shiftshare.Graph",
            new[] { "NodaTime" }));
}
