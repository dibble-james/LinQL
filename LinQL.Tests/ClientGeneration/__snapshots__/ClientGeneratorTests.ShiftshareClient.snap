#nullable enable
#pragma warning disable CS8618
namespace Shiftshare.Graph
{
    using System.Text.Json;
    using LinQL;
    using LinQL.Description;
    using LinQL.Translation;
    using Microsoft.Extensions.Logging;
    using Microsoft.Extensions.DependencyInjection;
    using NodaTime;

    [OperationType(RootOperationType.Query)]
    public partial class Query : RootType<Query>
    {
        public User? Me { get; set; }

        public IEvent[] Events { get; set; } = null !;
        [GraphQLOperation, GraphQLField(Name = "events")]
        public IEvent[] ExecuteEvents([GraphQLArgument(GQLType = "UUID!")] global::System.Guid calendar, [GraphQLArgument(GQLType = "OffsetDateTime!")] global::NodaTime.OffsetDateTime from, [GraphQLArgument(GQLType = "OffsetDateTime!")] global::NodaTime.OffsetDateTime to) => Events;
    }

    [OperationType(RootOperationType.Mutation)]
    public partial class Mutation : RootType<Mutation>
    {
        public User RegisterUser { get; set; }

        public IEvent CancelEvent { get; set; } = null !;
        [GraphQLOperation, GraphQLField(Name = "cancelEvent")]
        public IEvent ExecuteCancelEvent([GraphQLArgument(GQLType = "UUID!")] global::System.Guid calendarId, [GraphQLArgument(GQLType = "UUID!")] global::System.Guid id) => CancelEvent;
        public Calendar CreateCalendar { get; set; } = null !;
        [GraphQLOperation, GraphQLField(Name = "createCalendar")]
        public Calendar ExecuteCreateCalendar([GraphQLArgument(GQLType = "CreateCalendarInput!")] CreateCalendarInput input) => CreateCalendar;
        public IEvent RescheduleShift { get; set; } = null !;
        [GraphQLOperation, GraphQLField(Name = "rescheduleShift")]
        public IEvent ExecuteRescheduleShift([GraphQLArgument(GQLType = "UUID!")] global::System.Guid calendarId, [GraphQLArgument(GQLType = "UUID!")] global::System.Guid id, [GraphQLArgument(GQLType = "OffsetDate!")] global::NodaTime.OffsetDate date) => RescheduleShift;
        public IEvent ScheduleShift { get; set; } = null !;
        [GraphQLOperation, GraphQLField(Name = "scheduleShift")]
        public IEvent ExecuteScheduleShift([GraphQLArgument(GQLType = "UUID!")] global::System.Guid calendarId, [GraphQLArgument(GQLType = "UUID!")] global::System.Guid shiftType, [GraphQLArgument(GQLType = "OffsetDate!")] global::NodaTime.OffsetDate date) => ScheduleShift;
        public Calendar SetupShiftType { get; set; } = null !;
        [GraphQLOperation, GraphQLField(Name = "setupShiftType")]
        public Calendar ExecuteSetupShiftType([GraphQLArgument(GQLType = "UUID!")] global::System.Guid calendarId, [GraphQLArgument(GQLType = "ShiftTypeInput!")] ShiftTypeInput shiftType) => SetupShiftType;
        public ICalendarEvent[] SyncCalendarEvents { get; set; } = null !;
        [GraphQLOperation, GraphQLField(Name = "syncCalendarEvents")]
        public ICalendarEvent[] ExecuteSyncCalendarEvents([GraphQLArgument(GQLType = "UUID!")] global::System.Guid calendar, [GraphQLArgument(GQLType = "OffsetDateTime!")] global::NodaTime.OffsetDateTime after, [GraphQLArgument(GQLType = "CalendarEventInput!")] CalendarEventInput[] newEvents) => SyncCalendarEvents;
    }

    public enum ApplyPolicy
    {
        BEFORE_RESOLVER,
        AFTER_RESOLVER
    }

    public partial interface IEvent
    {
        public global::System.Guid Calendar { get; set; }

        public global::System.Guid Id { get; set; }

        public global::NodaTime.OffsetDate Date { get; set; }
    }

    public partial class Shift : IEvent
    {
        public IEvent Cancel { get; set; }

        public global::System.Guid ShiftType { get; set; }

        public global::System.Guid Calendar { get; set; }

        public global::System.Guid Id { get; set; }

        public int Version { get; set; }

        public global::System.Guid CorrelationId { get; set; }

        public string? State { get; set; }

        public global::NodaTime.OffsetDate Date { get; set; }

        public IEvent Reschedule { get; set; } = null !;
        [GraphQLOperation, GraphQLField(Name = "reschedule")]
        public IEvent ExecuteReschedule([GraphQLArgument(GQLType = "OffsetDate!")] global::NodaTime.OffsetDate to) => Reschedule;
    }

    public partial class User
    {
        public global::System.Guid Id { get; set; }

        public string Email { get; set; }

        public string Name { get; set; }

        public string Avatar { get; set; }

        public Calendar[] Calendars { get; set; }
    }

    public partial class Calendar
    {
        public global::System.Guid Id { get; set; }

        public string Name { get; set; }

        public CalendarAccess[] Access { get; set; }

        public ShiftType[] ShiftTypes { get; set; }

        public Shift ScheduleShift { get; set; } = null !;
        [GraphQLOperation, GraphQLField(Name = "scheduleShift")]
        public Shift ExecuteScheduleShift([GraphQLArgument(GQLType = "UUID!")] global::System.Guid shiftType, [GraphQLArgument(GQLType = "OffsetDate!")] global::NodaTime.OffsetDate date) => ScheduleShift;
    }

    public partial class CalendarAccess
    {
        public global::System.Guid User { get; set; }

        public PermissionAttributes Permission { get; set; }
    }

    public enum PermissionAttributes
    {
        OWNER
    }

    public partial class ShiftType
    {
        public global::System.Guid Id { get; set; }

        public string Name { get; set; }

        public global::NodaTime.OffsetTime Starts { get; set; }

        public global::NodaTime.Duration Duration { get; set; }
    }

    public partial class CreateCalendarInput
    {
        public global::System.Guid Id { get; set; }

        public string Name { get; set; }
    }

    public partial class ShiftTypeInput
    {
        public global::System.Guid Id { get; set; }

        public string Name { get; set; }

        public global::NodaTime.OffsetTime Starts { get; set; }

        public global::NodaTime.Duration Duration { get; set; }
    }

    public partial interface ICalendarEvent
    {
        public global::System.Guid Id { get; set; }

        public global::NodaTime.OffsetDateTime Raised { get; set; }

        public global::System.Guid Calendar { get; set; }

        public global::System.Guid EventId { get; set; }
    }

    public partial class ShiftScheduled : ICalendarEvent
    {
        public global::System.Guid ShiftType { get; set; }

        public global::NodaTime.OffsetDate Date { get; set; }

        public global::System.Guid Id { get; set; }

        public global::NodaTime.OffsetDateTime Raised { get; set; }

        public global::System.Guid Calendar { get; set; }

        public global::System.Guid EventId { get; set; }
    }

    public partial class ShiftRescheduled : ICalendarEvent
    {
        public global::NodaTime.OffsetDate Date { get; set; }

        public global::System.Guid Id { get; set; }

        public global::NodaTime.OffsetDateTime Raised { get; set; }

        public global::System.Guid Calendar { get; set; }

        public global::System.Guid EventId { get; set; }
    }

    public partial class EventCancelled : ICalendarEvent
    {
        public global::System.Guid Id { get; set; }

        public global::NodaTime.OffsetDateTime Raised { get; set; }

        public global::System.Guid Calendar { get; set; }

        public global::System.Guid EventId { get; set; }
    }

    public partial class CalendarEventInput
    {
        public ShiftScheduledInput? ShiftScheduled { get; set; }

        public ShiftRescheduledInput? ShiftRescheduled { get; set; }

        public EventCancelledInput? EventCancelled { get; set; }
    }

    public partial class ShiftScheduledInput
    {
        public global::System.Guid ShiftType { get; set; }

        public global::NodaTime.OffsetDate Date { get; set; }

        public global::System.Guid Id { get; set; }

        public global::NodaTime.OffsetDateTime Raised { get; set; }

        public global::System.Guid Calendar { get; set; }

        public global::System.Guid EventId { get; set; }
    }

    public partial class ShiftRescheduledInput
    {
        public global::NodaTime.OffsetDate Date { get; set; }

        public global::System.Guid Id { get; set; }

        public global::NodaTime.OffsetDateTime Raised { get; set; }

        public global::System.Guid Calendar { get; set; }

        public global::System.Guid EventId { get; set; }
    }

    public partial class EventCancelledInput
    {
        public global::System.Guid Id { get; set; }

        public global::NodaTime.OffsetDateTime Raised { get; set; }

        public global::System.Guid Calendar { get; set; }

        public global::System.Guid EventId { get; set; }
    }

    public static class LinQLOptionExtensions
    {
        public static LinQLOptions WithKnownScalars(this LinQLOptions options)
        {
            options.Scalars.Add(new("Int", "int"));
            options.Scalars.Add(new("Float", "float"));
            options.Scalars.Add(new("String", "string"));
            options.Scalars.Add(new("Id", "string"));
            options.Scalars.Add(new("Boolean", "bool"));
            options.Scalars.Add(new("UUID", "global::System.Guid"));
            options.Scalars.Add(new("Duration", "global::NodaTime.Duration"));
            options.Scalars.Add(new("OffsetDateTime", "global::NodaTime.OffsetDateTime"));
            options.Scalars.Add(new("OffsetDate", "global::NodaTime.OffsetDate"));
            options.Scalars.Add(new("OffsetTime", "global::NodaTime.OffsetTime"));
            return options;
        }
    }

    public static class JsonSerializerOptionsExtensions
    {
        public static JsonSerializerOptions WithKnownInterfaces(this JsonSerializerOptions options)
        {
            options.RegisterInterface<IEvent>(typeof(Shift));
            options.RegisterInterface<ICalendarEvent>(typeof(ShiftScheduled), typeof(ShiftRescheduled), typeof(EventCancelled));
            return options;
        }
    }
}
